File: src/lib/monitors.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/**
 * One row from `<project_root>/.cortex/monitors/monitors.json`. Mirrors the
 * Rust `MonitorSpec` struct in `src-tauri/src/monitors.rs`.
 */
export interface MonitorSpec {
  name: string;
  command: string;
  args: string[];
  level: "info" | "warn" | "error";
}

/**
 * Payload emitted on each line of monitor output. Mirrors the Rust
 * `MonitorLinePayload` struct in `src-tauri/src/monitors.rs`.
 */
export interface MonitorLinePayload {
  /** Monitor name, as configured in monitors.json. */
  name: string;
  /** One line of stdout or stderr from the child process. */
  line: string;
  /** Severity tag; stderr lines are bumped one notch from the spec default. */
  level: "info" | "warn" | "error";
  /** Unix epoch milliseconds when the line was forwarded. */
  ts: number;
}

/**
 * Start every monitor in `<project_root>/.cortex/monitors/monitors.json`.
 * Stops any previously running monitors first. Returns the names of the
 * monitors that were successfully spawned.
 */
export async function startMonitors(projectRoot: string): Promise<string[]> {
  return invoke<string[]>("start_monitors", { projectRoot });
}

/** Stop every running monitor. Idempotent. */
export async function stopMonitors(): Promise<void> {
  await invoke<void>("stop_monitors");
}

/**
 * Read `<project_root>/.cortex/monitors/monitors.json` without starting
 * anything. A missing file resolves to an empty list.
 */
export async function listMonitors(projectRoot: string): Promise<MonitorSpec[]> {
  return invoke<MonitorSpec[]>("list_monitors", { projectRoot });
}

/**
 * Subscribe to `monitor-line` events from the backend. The returned function
 * unsubscribes.
 */
export async function subscribeMonitorLines(
  cb: (payload: MonitorLinePayload) => void,
): Promise<UnlistenFn> {
  return listen<MonitorLinePayload>("monitor-line", (evt) => cb(evt.payload));
}

File: src-tauri/src/commands/monitors.rs
//! Tauri commands for the background monitor feature. See
//! `src-tauri/src/monitors.rs` for the underlying lifecycle module.
//!
//! # Wiring
//!
//! 1. `src-tauri/src/commands/mod.rs`: `pub mod monitors;`
//! 2. `src-tauri/src/lib.rs` `invoke_handler!`:
//!    - `commands::monitors::start_monitors`
//!    - `commands::monitors::stop_monitors`
//!    - `commands::monitors::list_monitors`
//! 3. `src-tauri/src/lib.rs` module tree: `pub mod monitors;`
//! 4. App shutdown: call `crate::monitors::stop_all().await` from a
//!    `RunEvent::ExitRequested` or `RunEvent::Exit` handler.

use std::path::PathBuf;

use crate::monitors;

/// Read `<project_root>/.cortex/monitors/monitors.json` and start every
/// configured monitor. Replaces any previously running monitors. Returns the
/// names of the monitors that were successfully spawned.
#[tauri::command]
pub async fn start_monitors(
    project_root: String,
    app: tauri::AppHandle,
) -> Result<Vec<String>, String> {
    let root = PathBuf::from(&project_root);
    if !root.is_dir() {
        return Err(format!(
            "start_monitors: project_root is not a directory: {project_root}"
        ));
    }
    monitors::start_all(&root, app)
        .await
        .map_err(|e| format!("failed to start monitors: {e:#}"))
}

/// Kill every running monitor. Idempotent.
#[tauri::command]
pub async fn stop_monitors() -> Result<(), String> {
    monitors::stop_all().await;
    Ok(())
}

/// Return the parsed contents of `<project_root>/.cortex/monitors/monitors.json`
/// without starting anything. Missing config ⇒ empty list (not an error).
#[tauri::command]
pub async fn list_monitors(project_root: String) -> Result<Vec<monitors::MonitorSpec>, String> {
    let root = PathBuf::from(&project_root);
    if !root.is_dir() {
        return Err(format!(
            "list_monitors: project_root is not a directory: {project_root}"
        ));
    }
    monitors::load_specs(&root).map_err(|e| format!("failed to list monitors: {e:#}"))
}
